# ECE-3 Lab 7

## Matrix Multiplication

### Exercise - 1

Write a function to compute the matrix product. The outline of the function `product_matrix(A, B)` is given below. Your result must be equivalent to `A@B` but you are not allowed to use any standard functions. 

- First check if computing matrix product is feasible.
- If possible, return a matrix equal to the product of the two input matrices. If not, return 0.
- Make sure the returned matrix is of proper shape.

In [ ]:
import numpy as np


def product_matrix(A, B):

  # Check if they can be multiplied
  shape_A = A.shape
  shape_B = B.shape
  # Check if columns of A are equal to rows of B
  if shape_A[1] != shape_B[0]:
    print('Cannot be multiplied')
    return 0

  # Compute the product
  # Initialize the output with zeros
  out = np.zeros((shape_A[0], shape_B[1]))
  # Loop-1 : Looping over rows of A
  for row_idx in range(shape_A[0]):    
    # Loop-2 : Looping over columns of B
    for col_idx in range(shape_B[1]):
      out[row_idx, col_idx] = np.dot(A[row_idx,:], B[:,col_idx])

  return out




# Do not edit below this line ###

print('----------\nTest Case: #1')
inpA = np.array([[1,2,3]])
inpB = np.array([[1],
                 [4],
                 [0]])

res1 = product_matrix(inpA, inpB).astype('int')
res = inpA @ inpB
print('Your output:\n', res1)
print('Expected output:\n', res)
try:
  assert(np.allclose(res1,res))
  print('Successful !')
except:
  print('Incorrect implementation. Please try again !')




print('\n\n---------\nTest Case: #2')
inpA = np.array([[1,2,3],
                 [3,4,5]])
inpB = np.array([[1,3],
                 [4,9],
                 [1,1]])

res1 = product_matrix(inpA, inpB).astype('int')
res = inpA @ inpB
print('Your output:\n', res1)
print('Expected output:\n', res)
try:
  assert(np.allclose(res1,res))
  print('Successful !')
except:
  print('Incorrect implementation. Please try again !')




print('\n\n---------\nTest Case: #3')
inpA = np.array([[1,2,3],
                 [3,4,5]])
inpB = np.array([[1,3],
                 [4,9]])

res1 = product_matrix(inpA, inpB)
res = 0
print('Your output:\n', res1)
print('Expected output:\n', res)
try:
  assert(np.allclose(res1,res))
  print('Successful !')
except AttributeError:
  print('Incorrect implementation. Please try again !')


----------
Test Case: #1
Your output:
 [[9]]
Expected output:
 [[9]]
Successful !


---------
Test Case: #2
Your output:
 [[12 24]
 [24 50]]
Expected output:
 [[12 24]
 [24 50]]
Successful !


---------
Test Case: #3
Cannot be multiplied
Your output:
 0
Expected output:
 0
Successful !


### Exercise - 2

Consider the matrices $\textbf{A}_{n\times m}$, $\textbf{B}_{m\times k}$, and $\textbf{C}_{k\times h}$. Dimensions (row $\times$ column) have been selected such that matrix products are computable.

\\

$\textbf{A} = \begin{bmatrix}
1 & 5 & 10 & 2 \\
2 & 0 & 3 & 23 \\
8 & 9 & 0 & 11
\end{bmatrix}$

$\textbf{B} = \begin{bmatrix}
10 & 6 \\
25 & 1 \\
13 & 0 \\
14& 11 
\end{bmatrix}$

$\textbf{C} = \begin{bmatrix}
5 & 10 & 7 & 8 \\
2 & 12 & 1 & 0
\end{bmatrix}$

\\

Compute the matrix product $\textbf{ABC}$. You may perform $\textbf{(AB)C}$ or $\textbf{A(BC)}$ - either works (due to the Associative Property). **You may input the matrices as numpy arrays and perform the linear algebra that way if you prefer.**

In [ ]:
# insert your matrix multiplication code below
A = np.array([[1, 5, 10, 2], [2, 0, 3, 23], [8, 9, 0, 11]])
B = np.array([[10, 6], [25, 1], [13, 0], [14, 11]])
C = np.array([[5, 10, 7, 8], [2, 12, 1, 0]])
# your result should match the answer given next
product_matrix(product_matrix(A,B),C)

array([[1531., 3326., 2084., 2344.],
       [2435., 6990., 2932., 3048.],
       [2651., 6726., 3391., 3672.]])



**Answer:** $\textbf{ABC} = \begin{bmatrix}
1531 & 3326 & 2084 & 2344 \\
2435 & 6990 & 2932 & 3048 \\
2651 & 6726 & 3391 & 3672
\end{bmatrix}$

### Exercise - 3

Consider the matrices $\textbf{A}_{nxm}$, $\textbf{B}_{mxk}$, and $\textbf{C}_{mxk}$. Dimensions have been selected so that the operation is computable. 

\\

$\textbf{A} = \begin{bmatrix}
1 & 5 & 10 & 2 \\
2 & 0 & 3 & 23 \\
8 & 9 & 0 & 11
\end{bmatrix}$

$\textbf{B} = \begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9 \\
10& 11& 12 
\end{bmatrix}$

$\textbf{C} = \begin{bmatrix}
13& 14& 15 \\
16& 17& 18 \\
19& 20& 21 \\
22& 23& 24 
\end{bmatrix}$

\\

Show the Distributive Property holds by computing $\textbf{A(B + C)}$ and $\textbf{AB + AC}$ and showing that they are equal, ie: $\textbf{A(B + C)} = \textbf{AB + AC}$. **You may input the matrices as numpy arrays and perform the linear algebra that way if you prefer.**

In [ ]:
# insert your code below for verifying the distributive property
A = np.array([[1, 5, 10, 2], [2, 0, 3, 23], [8, 9, 0, 11]])
B = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]])
C = np.array([[13, 14, 15], [16, 17, 18], [19, 20, 21], [22, 23, 24]])

# your result should match the answer given next
print(product_matrix(A,B+C) == (product_matrix(A,B) + product_matrix(A,C)))

[[ True  True  True]
 [ True  True  True]
 [ True  True  True]]




**Answer:** $\textbf{A(B+C)} = \begin{bmatrix}
438 & 474 & 510 \\
842 & 898 & 954 \\
644 & 700 & 756
\end{bmatrix}$

### Exercise - 4

Consider the matrices $\textbf{A}_{n\times m}$ and $\textbf{B}_{m\times k}$. Dimensions have been selected so that the operation is computable. 

$\textbf{A} = \begin{bmatrix}
1 & 3 & 5 & 7 & 9 \\
11& 13& 15& 17& 19
\end{bmatrix}$

$\textbf{B} = \begin{bmatrix}
2 & 4 & 6 & 8 \\
10& 12& 14& 16\\
18& 20& 22& 24\\
26& 28& 30& 32\\
34& 36& 38& 40
\end{bmatrix}$

\\

Show that the property $\textbf{(AB)}^T = \textbf{(B}^T\textbf{A}^T\textbf{)}$ holds by computing $\textbf{(AB)}^T$ and $\textbf{(B}^T\textbf{A}^T\textbf{)}$ and showing they are equal. **You may input the matrices as numpy arrays and perform the linear algebra that way if you prefer.**

In [ ]:
# insert your code below for verifying the above property
A = np.array([[1, 3, 5, 7, 9], [11, 13, 15, 17, 19]])
B = np.array([[2, 4, 6, 8], [10, 12, 14, 16], [18, 20, 22, 24], [26, 28, 30, 32], [34, 36, 38, 40]])

print(np.transpose(product_matrix(A, B)) == product_matrix(np.transpose(B), np.transpose(A)))

[[ True  True]
 [ True  True]
 [ True  True]
 [ True  True]]



**Answer:** $(\textbf{AB})^T = \begin{bmatrix}
610 & 1510\\
660 & 1660\\
710 & 1810\\
760 & 1960
\end{bmatrix}$

## Powers of a matrix

**Definition** For a positive integer $k$ , the $k^{th}$ power of a square matrix $A$ is defined as: $A^k=A.A.A....A\:\: (k$ times).

- By convention $A^0=I$, the identity matrix of the same dimension as $A$.
- It is **NOT** computed by raising the individual elements to the $k^{th}$ power.
- Use the operator `@` or `np.linalg.matrix_power()`.



In [ ]:
# Example for matrix power
import numpy as np
A = np.array([[4,1],
              [2,5]])
A_square = np.linalg.matrix_power(A,2)
print('A_square\n', A_square)
print('Shape: ', A_square.shape)

A_square_2 = A @ A
print('A_square_2\n', A_square_2)
print('Shape: ', A_square_2.shape)

A_square
 [[18  9]
 [18 27]]
Shape:  (2, 2)
A_square_2
 [[18  9]
 [18 27]]
Shape:  (2, 2)


## QR Factorization

**Proposition** Any matrix $A$ can be decomposed into two matrices $Q,R$ such that $A = QR$ and:
- $Q$ is a matrix such that $Q^TQ=I$ $\implies Q$ is orthogonal
- $R$ is a upper triangular matrix

Computing it in Python using `np.linalg.qr()`



In [ ]:
# Example for QR factorization

A = np.array([[4,1],
              [2,5]])
Q, R = np.linalg.qr(A)
print('Q\n', Q)
print('R\n', R)

# Verify that Q is an orthogonal matrix
print('Q is orthogonal')
print('Q.T @ Q\n',Q.T @ Q)
print('Q @ Q.T\n',Q @ Q.T)

Q
 [[-0.89442719 -0.4472136 ]
 [-0.4472136   0.89442719]]
R
 [[-4.47213595 -3.13049517]
 [ 0.          4.02492236]]
Q is orthogonal
Q.T @ Q
 [[ 1.00000000e+00 -6.81060746e-17]
 [-6.81060746e-17  1.00000000e+00]]
Q @ Q.T
 [[ 1.00000000e+00 -6.81060746e-17]
 [-6.81060746e-17  1.00000000e+00]]


### Exercise - 5

Write a function `is_orthogonal(A)` to check if a matrix `A` is orthogonal. The outline is given below. Return `True` if orthogonal else `False`.

In [ ]:

def is_orthogonal(A):
  #insert your code below
  # if orthogonal return True
  if np.mean(product_matrix(np.transpose(A), A) - np.eye(A.shape[0])) < 1e-10:
    return True
  else:
    return False
  
  # else return False









# Do not edit below this line ###

print('----------\nTest Case: #1')
A_temp = np.array([[1,10],
              [6,7]])
A,_ = np.linalg.qr(A_temp)
res = is_orthogonal(A)
res1 = True
try:
  assert(res==res1)
  print('Successful')
except:
  print('Incorrect implementation, please try again !')


print('----------\nTest Case: #2')
A = np.array([[1,0,0],
              [0,1,0],
              [0,0,1]])
res = is_orthogonal(A)
res1 = True
try:
  assert(res==res1)
  print('Successful')
except:
  print('Incorrect implementation, please try again !')



print('----------\nTest Case: #3')
A = np.array([[1,0,0],
              [0,1,1],
              [0,0,1]])
res = is_orthogonal(A)
res1 = False
try:
  assert(res==res1)
  print('Successful')
except:
  print('Incorrect implementation, please try again !')


----------
Test Case: #1
Successful
----------
Test Case: #2
Successful
----------
Test Case: #3
Successful
